### ISSUE TITLE
[tt-train] LoRA support

### ISSUE THREAD

**Issue by philei-tt** [tenstorrent/tt-metal#34782]
We need to add LoRA support to tt-train for fine-tuning. 

Ideally it should work ~1:1 like LoraConfig and LoraModel works in huggingface: https://huggingface.co/docs/peft/en/package_reference/lora 

LoraModel takes base model + LoraConfig
- It replaces all specifies Linear modules with LoraLinear 
- It freezes all weights in base model and keeps only LoraA and LoraB trainable

We might need some slight autograd updates to make updates easier

We should make sure that approach works well with both cpp and python modules

Some autograd optimizations are required to make LoRA actually useful: 
- Prune branches in autograd graph such that all nodes in a branch have requires_grad==False
- We can deallocate gradient tensor as soon as it's propagated as long as it's activation and not a weight. 
- TODO: Any other ideas?
